cekit/descriptor/image.py

```diff
@@ -45,48 +50,318 @@ def __init__(self, descriptor, artifact_dir):
-        self['labels'].extend([{'name': 'org.concrt.version',
-                                'value': cekit_version},
-                               {'name': 'io.cekit.version',
-                                'value': cekit_version}])
-
-        # The description key available in image descriptor's
-        # root is added as labels to the image
-        key = 'description'
-
-        # If we define the label in the image descriptor
-        # we should *not* override it with value from
-        # the root's key
-        if key in self._descriptor and not self.label(key):
-            value = self._descriptor[key]
-            self._descriptor['labels'].append({'name': key, 'value': value})
-
-        # Last - if there is no 'summary' label added to image descriptor
-        # we should use the value of the 'description' key and create
-        # a 'summary' label with it's content. If there is even that
-        # key missing - we should not add anything.
-        description = self.label('description')
-
-        if not self.label('summary') and description:
-            self._descriptor['labels'].append(
-                {'name': 'summary', 'value': description['value']})
-
         self._descriptor['labels'] = [Label(x) for x in self._descriptor.get('labels', [])]
         self._descriptor['envs'] = [Env(x) for x in self._descriptor.get('envs', [])]
         self._descriptor['ports'] = [Port(x) for x in self._descriptor.get('ports', [])]
         if 'run' in self._descriptor:
             self._descriptor['run'] = Run(self._descriptor['run'])
         self._descriptor['artifacts'] = [Resource(a, directory=self._artifact_dir)
                                          for a in self._descriptor.get('artifacts', [])]
-        if 'modules' in self._descriptor:
-            self._descriptor['modules'] = Modules(self._descriptor['modules'], self.path)
-        if 'packages' in self._descriptor:
-            self._descriptor['packages'] = Packages(self._descriptor['packages'])
+        self._descriptor['modules'] = Modules(self._descriptor.get('modules', {}), self.path)
+        self._descriptor['packages'] = Packages(self._descriptor.get('packages', {}))
         if 'osbs' in self._descriptor:
             self._descriptor['osbs'] = Osbs(self._descriptor['osbs'])
         self._descriptor['volumes'] = [Volume(x) for x in self._descriptor.get('volumes', [])]
+
+        # make sure image declarations override any module definitions
+        self._image_overrides = { 'artifacts': self._to_dict(self.artifacts), 'modules': self._to_dict(self.modules.install) }
+        self._all_artifacts = self._to_dict(self.artifacts)
+
+    def process_defaults(self):
+        """Prepares default values before rendering"""
+        if not self.run:
+            self.run = Run({})
+
+        # do we want to force a user?
+        if 'user' not in self.run:
+            self.run._descriptor['user'] = cekit.DEFAULT_USER
+
+    @property
```

Comment:
If we want to go for properties, we should do it for all the descriptor objects and remove collections.MutableMapping from object hierarchy and consider removing __getitem__, __delitem__ and other dunder methods so we have only one and consistent way of accessing self._dictionary. This PR opens a way of inconsistency and is error prone as we can have different logic for accessing inner dictionary.

Reply:
I can look at doing this as a follow on effort, but it's a bit of work and most likely error prone.

Reply:
its error prone, but its better to refactor and break things now than later after more testing.

Reply:
ok. I'm changing review to OK, but I'll suggest to refactor code to new @properties like access as code should be consistent. @goldmann any opinions?